File: mobile/lib/agent/style_draw.dart
import 'dart:math' as math;

String drawStyleTags(List<String> tags, List<String> pinned, int count,
    double min, double max, int seed) {
  var state = seed & 0xffffffff;
  if (state == 0) state = 0x6d2b79f5;
  double random() {
    state = (state ^ (state << 13)) & 0xffffffff;
    state = (state ^ (state >> 17)) & 0xffffffff;
    state = (state ^ (state << 5)) & 0xffffffff;
    return state / 4294967296;
  }

  final fixed = pinned.where(tags.contains).toSet().toList();
  final pool = tags.toSet().where((t) => !fixed.contains(t)).toList();
  final chosen = [...fixed];
  for (var i = 0; i < math.max(0, math.min(pool.length, count)); i++) {
    final at = i + (random() * (pool.length - i)).floor();
    final t = pool[i];
    pool[i] = pool[at];
    pool[at] = t;
    chosen.add(pool[i]);
  }
  final low = math.max(0.0, math.min(3.0, math.min(min, max)));
  final high = math.max(low, math.min(3.0, math.max(min, max)));
  return chosen.map((tag) {
    final weight = ((low + random() * (high - low)) * 100).round() / 100;
    return '${weight.toStringAsFixed(2).replaceFirst(RegExp(r'\.?0+$'), '')}::$tag::';
  }).join(', ');
}

String appendStylePrompt(String base, String addition) {
  if (addition.trim().isEmpty) return base;
  return base.trim().isEmpty
      ? addition
      : '$base${base.trimRight().endsWith(',') ? ' ' : ', '}$addition';
}

File: src/tavern/style-draw.ts
export function drawStyleTags(tags: string[], pinned: string[], count: number, min: number, max: number, seed: number): string {
  let state = (seed >>> 0) || 0x6d2b79f5;
  const random = () => { state ^= state << 13; state ^= state >>> 17; state ^= state << 5; return (state >>> 0) / 4294967296; };
  const fixed = [...new Set(pinned.filter((tag) => tags.includes(tag)))];
  const pool = [...new Set(tags)].filter((tag) => !fixed.includes(tag));
  const chosen = [...fixed];
  for (let i = 0; i < Math.max(0, Math.min(pool.length, Math.trunc(count))); i++) {
    const at = i + Math.floor(random() * (pool.length - i));
    [pool[i], pool[at]] = [pool[at], pool[i]];
    chosen.push(pool[i]);
  }
  const low = Math.max(0, Math.min(3, Math.min(min, max)));
  const high = Math.max(low, Math.min(3, Math.max(min, max)));
  return chosen.map((tag) => `${(Math.round((low + random() * (high - low)) * 100) / 100).toFixed(2).replace(/\.?0+$/, "")}::${tag}::`).join(", ");
}
export function appendStylePrompt(base: string, addition: string): string {
  if (!addition.trim()) return base;
  return base.trim() ? base + (base.trimEnd().endsWith(",") ? " " : ", ") + addition : addition;
}
